File: apps/backend/src/main/java/com/digisphere/pontofacil/service/CompanyService.java
package com.digisphere.pontofacil.service;

import com.digisphere.pontofacil.domain.company.Company;
import com.digisphere.pontofacil.dtos.company.CompanyDTO;
import com.digisphere.pontofacil.repository.CompanyRepository;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.stream.Stream;

@Service
@Transactional
public class CompanyService {

  private final CompanyRepository companyRepository;

  public CompanyService(CompanyRepository companyRepository) {
    this.companyRepository = companyRepository;
  }

  public Optional<CompanyDTO> save(CompanyDTO dto) {
    final var entity = dto.toEntity();

    return Optional.of(
      companyRepository
        .save(entity)
        .toDto()
    );
  }

  public Optional<CompanyDTO> getById(Long id) {
    return Optional.of(companyRepository.findById(id)
      .map(Company::toDto)
      .orElseThrow());
  }

  public Stream<CompanyDTO> getAll() {
    return companyRepository.findAll()
      .stream()
      .map(Company::toDto);
  }


}

File: apps/backend/src/main/kotlin/com/digisphere/pontofacil/company/service/CompanyService.kt
package com.digisphere.pontofacil.company.service

import com.digisphere.pontofacil.company.api.CompanyDTO
import com.digisphere.pontofacil.company.repository.CompanyRepository
import com.digisphere.pontofacil.company.toDTO
import com.digisphere.pontofacil.company.toEntity
import org.slf4j.LoggerFactory
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class CompanyService(private val repository: CompanyRepository) {

  private val logger = LoggerFactory.getLogger(javaClass)

  fun create(dto: CompanyDTO): Result<CompanyDTO> = runCatching {
    repository.save(dto.toEntity()).toDTO()
  }.onFailure {
    logger.error("Unexpected error while creating company", it)
  }

  @Transactional(readOnly = true)
  fun findById(id: Long): Result<CompanyDTO> = runCatching {

    val company = repository.findByIdOrNull(id)
    requireNotNull(company) { "Company not found with id: $id" }
    company.toDTO()

  }.onFailure {
    logger.error("Unexpected error while finding company by id: $id", it)
  }

  @Transactional(readOnly = true)
  fun findAll(): Result<List<CompanyDTO>> = runCatching {
    repository.findAll().map { it.toDTO() }
  }.onFailure {
    logger.error("Unexpected error while fetching all companies", it)
  }
}
